Bugfix/reduce memusage plugins

TD-er opened the pull request:
Also introduction of `Plugin_task_data` array.
This array (see `_Plugin_helper.h` ) is there to get a global array of member data used in any plugin.
This allows multiple instances of a plugin, since the key to store/retrieve the data is the task index.

jimmys01 committed:
Merge pull request #3 from letscontrolit/mega

Merge it

TD-er committed:
[Nextion] Reduce memory usage when not active

TD-er committed:
[Nextion] Reduce memory usage when not active

TD-er committed:
[p073_7DGT] Reduce memory usage of 7-segment display plugin

Also removed lots and lots of duplicate code and improved readability.

TD-er committed:
[p073_7DGT] Reduce memory usage of 7-segment display plugin

Also removed lots and lots of duplicate code and improved readability.

jimmys01 committed:
Merge pull request #4 from letscontrolit/mega

test

jimmys01 committed:
Bump up IrRemote Lib to 2.5.4

jimmys01 committed:
Add the newest IR protocols to EspEasy

TD-er committed:
[memory] Reduce memory usage of p070_NeoPixelClock when not enabled

TD-er committed:
[memory] Reduce memory usage of p070_NeoPixelClock when not enabled

TD-er committed:
[memory] Reduce memory of P082_GPS and P068_SHT3x when not enabled

Also introduce a global pointer array to keep plugin settings per task ID to allow multiple instances of the same plugin.

TD-er committed:
[memory] Reduce memory of P082_GPS and P068_SHT3x when not enabled

Also introduce a global pointer array to keep plugin settings per task ID to allow multiple instances of the same plugin.

TD-er committed:
[memory] P077_CSE7766 Use Plugin_task_data array

TD-er committed:
[memory] P077_CSE7766 Use Plugin_task_data array

TD-er commented:
Would be nice if these changed plugins could be tested.
Since there was already a report of Nextion plugin not working since the HW serial selector, I guess it is still broken. (see #2228 )

P077 and the P082_GPS plugin are already being tested here on my own hardware.
P073 is a major rewrite to remove a lot of code duplication, so that one should be tested for sure.

GPS plugin should now be usable when no GPS is connected (not tested myself, just looked at the code and found a missing check for `isInitialized()`.

Barracuda09 commented on src/_P077_CSE7766.ino:
```diff
@@ -380,8 +376,10 @@ boolean Plugin_077(byte function, struct EventStruct *event, String &string) {
 bool CseReceived(struct EventStruct *event) {
-  if (!P077_data)
+  P077_data_struct* P077_data = static_cast<P077_data_struct*>(getPluginTaskData(event->TaskIndex));
+  if (!P077_data) {
```
Maybe do something like this for pointers, so you don't get confused with bool comparison:
```
if (P077_data == nullptr) {

Barracuda09 replied:
There are more situation were this is used to compare pointers with bool type compare.

TD-er replied:
That's true.
Also it's a plus the compiler will then give a warning or more likely an error.
I will change it.
Thanks for noticing.

TD-er committed:
[memory] P075_Nextion Use Plugin_task_data array

TD-er committed:
[memory] P075_Nextion Use Plugin_task_data array

TD-er committed:
[memory] P074_TSL2591 Use Plugin_task_data array

TD-er committed:
[memory] P074_TSL2591 Use Plugin_task_data array

TD-er committed:
[memory] P073_7DGT Use Plugin_task_data array

TD-er committed:
[memory] P073_7DGT Use Plugin_task_data array

TD-er committed:
[memory] P070_NeoPixel_Clock Use Plugin_task_data array

TD-er committed:
[memory] P070_NeoPixel_Clock Use Plugin_task_data array

TD-er committed:
[memory] P046_VentusW266 Use member data struct

Not using the Plugin_task_data, since it is using interrupts and callback functions.
Not sure what will happen if we store these volatile variables in the same global array.

TD-er committed:
[memory] P046_VentusW266 Use member data struct

Not using the Plugin_task_data, since it is using interrupts and callback functions.
Not sure what will happen if we store these volatile variables in the same global array.

TD-er committed:
[memory] P045_MPU6050 Use Plugin_task_data array

TD-er committed:
[memory] P045_MPU6050 Use Plugin_task_data array

Barracuda09 reviewed (changes requested):
Maybe do something like this for pointers, so you don't get confused with bool comparison:

```
if (P077_data == nullptr) {

jimmys01 committed:
[IRTX] Even more timming Buffer added

I keep bumping to AC units with even bigger appetite for lengthy timings. Hopefully this time will be enough

TD-er committed:
[memory] Change pointer checks to compare with nullptr

As suggested [here](https://github.com/Barracuda09)

TD-er committed:
[memory] Change pointer checks to compare with nullptr

As suggested [here](https://github.com/Barracuda09)

TD-er committed:
[memory] P044_P1WiFiGateway Use Plugin_task_data array and String as buffer

The internal buffer was a char* array, but all it needed was just the same as being already available in the `String` class.
So it has been changed to be a String.

TD-er committed:
[memory] P044_P1WiFiGateway Use Plugin_task_data array and String as buffer

The internal buffer was a char* array, but all it needed was just the same as being already available in the `String` class.
So it has been changed to be a String.

TD-er commented:
The P044_P1WiFiGateway plugin is now added and it needed quite a big change to use String instead of the char*.
So it really needs to be tested before merging.

Barracuda09 commented:
> So it really needs to be tested before merging.

I can see what I can make of it... NOOB in PlatformIO.

TD-er commented:
@Barracuda09 I will make a test build for it. (build running as we speak)

TD-er commented:
And here is a [testbuild](https://www.dropbox.com/s/rdayarjsf0bkiji/ESPEasy_mega-20190110-14-PR_2235.zip?dl=0)

jimmys01 committed:
[IRTX] Allocate big buffers on the heap for RAW and RAW2 encodings

jimmys01 committed:
Update _P035_IRTX.ino

Barracuda09 commented:
@TD-er I can not test all aspects of these changes because I only have DSMR 2.2 (So no checksum).

I still have to check it by the way

TD-er commented:
As long as it is still running the basics of that plugin, then I'm convinced the rest will work too.

TD-er committed:
[memory] Shared code to select Serial Config (8N1 etc)

Also fixed a minor issue in GPS plugin destructor

TD-er committed:
[memory] Shared code to select Serial Config (8N1 etc)

Also fixed a minor issue in GPS plugin destructor

TD-er committed:
[memory] P020_Ser2Net Use Plugin_task_data array

TD-er committed:
[memory] P020_Ser2Net Use Plugin_task_data array

TD-er committed:
[ESP32] Change P002_ADC range to 4095 for ESP32

See https://github.com/letscontrolit/ESPEasy/issues/1632#issuecomment-454952002

TD-er committed:
[ESP32] Change P002_ADC range to 4095 for ESP32

See https://github.com/letscontrolit/ESPEasy/issues/1632#issuecomment-454952002

TD-er committed:
[Settings] Added check for offset on last member of struct

An extra check to see if building on different platforms makes a difference in offset within the SettingsStruct.

TD-er committed:
[Settings] Added check for offset on last member of struct

An extra check to see if building on different platforms makes a difference in offset within the SettingsStruct.

TD-er commented:
Did anybody test these changes?

clumsy-stefan commented:
I have this PR included in all my latest builds (which run for >24hr now) but obviously I did/could not test all plugins individually... only thing I found was that 
https://github.com/letscontrolit/ESPEasy/pull/2235/commits/b7782944a6894fdb05578d3ab16b763a4d585a11 tells me now, that there is an offset error in my settings, but I use 24 tasks in my builds, so I assume that's why....

TD-er commented:
Yep, if you're running 24 tasks, then you will have an offset.
We could make the 24-tasks config usable via a define, then also this check can be made dynamic and maybe also alert to discard the settings when settings are loaded with another number of tasks set.

By the way, which plugins did you test?

clumsy-stefan commented:
More or less since the beginning I use the following defines which I found somewhere in a WiKi or Forum:
```
#define TASKS_MAX                          24 // max 12! 
#define DAT_OFFSET_CONTROLLER            DAT_OFFSET_TASKS + (DAT_TASKS_DISTANCE * TASKS_MAX)                        // each controller = 1k, 4 max 
#define DAT_OFFSET_CUSTOM_CONTROLLER     DAT_OFFSET_CONTROLLER + (DAT_CUSTOM_CONTROLLER_SIZE * CONTROLLER_MAX)  // each custom controller config = 1k, 4 max 
```
this seems to work out fine. I use 24 takss for 4M and 16M builds and 12 tasks for 1M builds.
probably these could be used generally and also checked for?

On my 4M and 16M nodes I use the plugins 1, 2, 5, 6, 7, 9, 15, 19, 22, 25, 26, 28, 33, 36, 68, 69, 72, 77, 81, 124, 130 and 133.
On my 1M nodes I only use 1 and 26.
And on all of them only controller 9.

TD-er commented:
If you run the command "meminfodetail" and then go to the sysinfo page, you'll see a drawing of the settings layout at the bottom of the page.
Can you give a screenshot here of that drawing?

Edit:
The only plugin you use which has also changed for this PR is 77.
So would be nice if others could also be checked.
The are: 20, 44, 45, 46, 70, 73, 74, 75, 77, 82

clumsy-stefan commented:
Screenshot attached:
<img width="800" alt="image" src="https://user-images.githubusercontent.com/32835100/51462478-00904980-1d61-11e9-991e-acf27fe8d000.png">
<img width="787" alt="image" src="https://user-images.githubusercontent.com/32835100/51462504-13a31980-1d61-11e9-9af0-18a6d51cd950.png">
<img width="787" alt="image" src="https://user-images.githubusercontent.com/32835100/51462534-1f8edb80-1d61-11e9-8e1f-2058654c1cb9.png">

jimmys01 committed:
[Controllers] Fix small error in commit #2246 

Just realized that I had mixed up the default defines. It is now OK.

clumsy-stefan committed:
[1M minimal Size] Undefine graphical meminfodetail

Undefine graphical meminfodetail if BUILD_MINIMAL_OTA is defined to reduce sketch size.
Reduces sketch about 3k.

clumsy-stefan committed:
[1M minimal Size] Exclude all debug logs from standard files

Exclude all >= LOG_LEVEL_DEBUG except in Controllers, Plugins and Notifications.
Reduces sketch about 3k.

TD-er commented:
A new [test build](https://www.dropbox.com/s/v66yth2fokaeyra/ESPEasy_mega-20190121-18-PR_2235.zip?dl=0)

TD-er committed:
[TinyGPS++] Added GSA and GSV sentences to get satellite ID & SNR

TD-er committed:
[TinyGPS++] Added GSA and GSV sentences to get satellite ID & SNR

TD-er committed:
[GPS] Use GSV and GSA sentences to view satellite ID and SNR

TD-er committed:
[GPS] Use GSV and GSA sentences to view satellite ID and SNR

TD-er committed:
[PlatformIO] Allow building ESP32 espressif32@1.6.0

TD-er committed:
[PlatformIO] Allow building ESP32 espressif32@1.6.0

TD-er committed:
[PubSubClient] Increase MQTT_MAX_PACKET_SIZE from 384 to 1024

TD-er committed:
[PubSubClient] Increase MQTT_MAX_PACKET_SIZE from 384 to 1024

TD-er commented:
esp32dev build fails, due to sketch size exceeding default partition.
That's why I already made a 1M8 partition for the ESP32 sketch partitions.
So I guess we should abandon that build?

Grovkillen committed:
Merge pull request #7 from letscontrolit/mega

update from master

Grovkillen committed:
[flasher] fixed empty folder issue

clumsy-stefan committed:
[1M minimal Size] Include BUILD_NO_DEBUG in plugin sets

Add BUILD_NO_DEBUG in define_plugin_sets.h to PLUGIN_BUILD_MINIMAL_OTA.
This makes commit #316691ad07731798687ada096907c5ccb61fb1b6 included and active in 1M minimal OTA build.

clumsy-stefan committed:
[1M minimal Size] Remove unused variable..

...and make travis compiler happy again!

clumsy-stefan committed:
[1M minimal Size] Remove another unused variable..

I have no clue why my compiler does not complain about these, even though I have -Werror=unused-variable in the command line options...

clumsy-stefan committed:
[1M minimal Size] Exclude debug from minimal Controllers/Plugins

Exclude debug logs from C001, C002, C005, C008, C009, C013 and P001, P026, P033 which are compiled  in by default in the PLUGIN_BUILD_MINIMAL_OTA set.
Reduces sketch about 1k.

TD-er committed:
[GPS] Improve displayed satellite stats

TD-er committed:
[GPS] Improve displayed satellite stats

TD-er committed:
[WiFi] Allow force to B/G to improve WiFi stability.

See #2012 and discussion here: https://github.com/letscontrolit/ESPEasy/issues/1987#issuecomment-455843832

TD-er committed:
[WiFi] Allow force to B/G to improve WiFi stability.

See #2012 and discussion here: https://github.com/letscontrolit/ESPEasy/issues/1987#issuecomment-455843832

TD-er committed:
[WiFi] Correct fall-back when B/G only mode cannot connect

TD-er committed:
[WiFi] Correct fall-back when B/G only mode cannot connect

TD-er committed:
Merge branch 'mega' into bugfix/reduce_memusage_plugins

jimmys01 committed:
Merge branch 'mega' into mega

TD-er committed:
Moved testing plugins to test builds & added safety docs for power sensors

See: https://github.com/letscontrolit/ESPEasy/issues/2266

TD-er committed:
Moved testing plugins to test builds & added safety docs for power sensors

See: https://github.com/letscontrolit/ESPEasy/issues/2266

TD-er committed:
Merge remote-tracking branch 'origin/bugfix/reduce_memusage_plugins' into bugfix/reduce_memusage_plugins

TD-er committed:
[WiFi] Add option to set wifi off and restart wifi when connection lost

As discussed here: https://github.com/letscontrolit/ESPEasy/issues/1987#issuecomment-457268319
In short, this new setting allows to set wifi completely off when a disconnect is being handled.
This will force a complete restart and configure of the wifi setup.

TD-er committed:
[WiFi] Add option to set wifi off and restart wifi when connection lost

As discussed here: https://github.com/letscontrolit/ESPEasy/issues/1987#issuecomment-457268319
In short, this new setting allows to set wifi completely off when a disconnect is being handled.
This will force a complete restart and configure of the wifi setup.

TD-er commented:
Made a [test build](https://www.dropbox.com/s/izlb4xjjkrgcx01/ESPEasy_mega-20190121-29-PR_2235.zip?dl=0) of all commits upto the option to switch wifi off when trying to reconnect.

Would be nice if these can be tested and I think I will merge this PR tomorrow if nothing really bad is reported when running this build.

clumsy-stefan commented:
I also made a build and installed it on my debug-nodes. one with wifi_off option enabled, the other one without. both initially connected in N mode... I'll see what happens (when I tried in my tsts with wifi_off it did not help... let's see...)

all other commits (including the force g mode from yesterday) ran fine for a couple of days on my nodes though (obviously the last one only since yesterday on 2 test nodes)...

clumsy-stefan commented:
@TD-er on the testnode with "Restart WiFi on lost conn.:" enabled and "Force B/G" disabled I already had 3 reboots, always the same behaviour after the 4way HS timeout always Auth timeouts until an exception raises... 
```
1453113 : WIFI : Disconnected! Reason: '(16) Group key update timeout' Connected for 24 m 9 s
1453114 : WIFI : Switch off WiFi
1453350 : EVENT: WiFi#Disconnected
1453408 : WIFI : Disconnected! Reason: '(202) Auth fail' Connected for 24 m 9 s
1453534 : WIFI : Set WiFi to STA
1453566 : WIFI : Connecting clumsy_ap2 attempt #0
1453570 : WIFI : Not configured in Station Mode!!: clumsy_ap2
1455681 : EVENT: WiFi#Disconnected
1455738 : WIFI : Disconnected! Reason: '(2) Auth expire' Connected for 2012 ms
1455739 : WIFI : Switch off WiFi
1455881 : WIFI : Set WiFi to STA
1455913 : WIFI : Connecting clumsy_ap2 attempt #1
1455916 : WIFI : Not configured in Station Mode!!: clumsy_ap2
1459880 : EVENT: WiFi#Disconnected
1459949 : WIFI : Disconnected! Reason: '(2) Auth expire' Connected for 3863 ms
1459950 : WIFI : Switch off WiFi
1460088 : WIFI : Set WiFi to STA
```
until
```
1487212 : WIFI : Connecting clumsy_ap2 attempt #8
1487215 : WIFI : Not configured in Station Mode!!: clumsy_ap2
1491179 : EVENT: WiFi#Disconnected
1491263 : WIFI : Disconnected! Reason: '(2) Auth expire' Connected for 3862 ms
1491266 : WIFI : Switch off WiFi
1491877 : WIFI : Set WiFi to STA
1491911 : WIFI : Set WiFi to AP+STA
1491944 : WIFI : [AP] softAPConfig failed!
1491947 : WIFI : Error while starting AP Mode with SSID: wemos-mini-17_17 IP: 192.168.4.1
1491949 : WIFI : Connecting clumsy_ap2 attempt #9
1491953 : WIFI : Not configured in Station Mode!!: clumsy_ap2
1495913 : EVENT: WiFi#Disconnected

Exception (28):
epc1=0x4025fbe6 epc2=0x00000000 epc3=0x00000000 excvaddr=0x00000000 depc=0x00000000
```
but no reconnection!
 I'll do a bit further test with "Force B/G" still...

EDIT: it never reaches the "fallback to B/G" as it crashes before the 10th try... blank node only sysinfo and fhem controller enabled...

TD-er commented:
Maybe we also have to look into actively closing all socket connections and restarting them after the connection is rebuilt.
Like I said, I am not entirely sure this is properly done.

clumsy-stefan commented:
sure, just wanted to give feedback that this commit does not seem to change anything...

clumsy-stefan commented:
it also seems that the "fallback to G" does not work, I have my two test nodes with over 300 reconnect tries:
```
47522244 : WD   : Uptime 792 ConnectFailures 6 FreeMem 16184 WiFiStatus 0
47524666 : WIFI : Set WiFi to AP+STA
47524707 : EVENT: WiFi#APmodeEnabled
47524742 : WIFI : AP Mode ssid will be wemos-mini-18_18 with address 192.168.4.1
47524744 : WIFI : Connecting clumsy_ap2 attempt #527
47524747 : WIFI : Not configured in Station Mode!!: clumsy_ap2
47530012 : EVENT: WiFi#Disconnected
47530048 : WIFI : Disconnected! Reason: '(15) 4way handshake timeout' Connected for 5128 ms
```
and then
```
47631801 : WIFI : Connecting clumsy_ap2 attempt #528
47631803 : WIFI : Not configured in Station Mode!!: clumsy_ap2
47633127 : EVENT: WiFi#Disconnected
47633163 : WIFI : Disconnected! Reason: '(2) Auth expire' Connected for 1122 ms
```
and so on...

clumsy-stefan commented:
did some further tests, and (to me) it seems that changing the mode (from N to B/G) while/after a succefull connection is/was made is not working... it only connects to B/G when it's done after booting... (see log above)... but that's just MHO...

TD-er commented:
In my tests, I performed a WiFi disconnect command and then it reconnects using G.
It doesn't kill the existing connection if you change the setting, since it is only used while establishing the connection.

clumsy-stefan commented:
yes, if you actively do a disconnct or reboot... if it runs into the "auth expire" iisue as above it won't change it (I think)....

TD-er committed:
Merge pull request #2232 from jimmys01/mega

[IRTX] [IRRX] Update IrRemote Library and corresponding plugins

TD-er committed:
Merge pull request #2263 from Grovkillen/UpdateDocs

[flasher] fixed "empty folder" issue #2247

TD-er committed:
Merge pull request #2262 from clumsy-stefan/reduce_1M_minimal_ota

[Sketch size] Reduce sketch size if needed

TD-er committed:
Merge pull request #2261 from jimmys01/patch-1

[Controllers] Fix small error in commit #2246

TD-er committed:
Merge remote-tracking branch 'origin/bugfix/reduce_memusage_plugins' into bugfix/reduce_memusage_plugins

TD-er commented:
Did anyone test the changed plugins in this PR?

TD-er committed:
[#2254] Reduce memory usage on MH-Z19 and improve reading stability

Reading the sensor data will now only accept data starting with 0xFF.
This should improve reading stability.

Also added indicator showing checksum pass/fail and the model detected (MH-Z19 A or B)

TD-er committed:
[#2254] Make it compile in all core libraries

Linker issue with F() macro.

TD-er committed:
[ESP8266] Do not use default for TX pin

The RX/TX pins are defined in the init. So no need to hand them to the begin function call again and set some default when not used.
This will lead to issues where the TX pin is not working when using HW serial.

TD-er committed:
[Core 2.5.0] Move tag to beta3

See commit: https://github.com/esp8266/Arduino/commit/21db8fc9285d2f548c0123206d53cab059cf3b3d

TD-er committed:
[SenseAir] Improvement on Modbus handling

As a preparation for a generic Modbus layer for other plugins too.

TD-er committed:
[SenseAir] Some more improvements Modbus implementation

TD-er committed:
[WiFi] Destruct WiFi object when wifi needs to be reset

See https://github.com/esp8266/Arduino/issues/5527#issuecomment-460537616

TD-er committed:
[Eastron] Add CRC pass/fail stats to plugin page and allow retries

TD-er committed:
[Eastron] Fix compile error

TD-er committed:
[#2292] HX711: tareChanA command is not case insensitive

TD-er committed:
[ESPeasySerial] return -1 as pins on unset serial type

TD-er committed:
[GPS] Fix some uninitialized issues

TD-er committed:
[PlatformIO] Core 2.5.0 is no longer beta

Regular platform is still set to 2.4.2, since 2.5.0 causes issues.

TD-er committed:
[SWserial] Set TX interrupts for 9600 baud

This was the used setting in older versions of SoftwareSerial

TD-er committed:
[Build] Rename build files and disable core 2.5.0 builds

Core 2.5.0 is out, but this gives instability issues serving web pages.
For now the builds of core 2.5.0 are not included in the nightly builds until this is fixed.

TD-er committed:
[ESPeasySerial] Make it compile for core 2.4.1

TD-er committed:
[Serial wrapper] Fix serial issues (for ESP8266) when using HW serial
